packages/file-download/src/browser/file-download-command-contribution.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable } from 'inversify';
+import URI from '@theia/core/lib/common/uri';
+import { notEmpty } from '@theia/core/lib/common/objects';
+import { UriSelection } from '@theia/core/lib/common/selection';
+import { SelectionService } from '@theia/core/lib/common/selection-service';
+import { Command, CommandContribution, CommandRegistry } from '@theia/core/lib/common/command';
+import { UriAwareCommandHandler, UriCommandHandler } from '@theia/core/lib/common/uri-command-handler';
+import { FileDownloadService } from './file-download-service';
+
+@injectable()
+export class FileDownloadCommandContribution implements CommandContribution {
+
+    @inject(FileDownloadService)
+    protected readonly downloadService: FileDownloadService;
+
+    @inject(SelectionService)
+    protected readonly selectionService: SelectionService;
+
+    registerCommands(registry: CommandRegistry): void {
+        const options = {
+            multi: true,
+            isValid: this.isValid.bind(this)
+        };
+        const handler = new UriAwareCommandHandler<URI[]>(this.selectionService, this.downloadHandler(), options);
+        registry.registerCommand(FileDownloadCommands.DOWNLOAD, handler);
+    }
+
+    protected downloadHandler(): UriCommandHandler<URI[]> {
+        return {
+            execute: uri => this.executeDownload(uri),
+            isEnabled: (uri, args) => this.isDownloadEnabled(uri, args),
+            isVisible: (uri, args) => this.isDownloadVisible(uri, args),
+        };
+    }
+
+    protected async executeDownload(uris: URI[]): Promise<void> {
+        this.downloadService.download(uris);
+    }
+
+    // tslint:disable-next-line:no-any
+    protected isDownloadEnabled(uri: Object | undefined, ...args: any[]): boolean {
+        return this.getUris(uri).length > 0;
+    }
+
+    // tslint:disable-next-line:no-any
+    protected isDownloadVisible(uri: Object | undefined, ...args: any[]): boolean {
+        return this.isDownloadEnabled(uri, args);
+    }
+
+    protected isValid(uris: URI[]): boolean {
+        if (uris.length === 0) {
+            return false;
+        }
+        if (uris.length === 1) {
+            return true;
+        }
+        // Can download multiple files iff they are from the same container folder.
```

Comment:
Why restrict it like that?

Reply:
I am OK removing the restriction if you could help me define a clear way how the files should be zipped up in a folder when selecting multiple files "randomly" from a workspace.

Reply:
We could preserve the structure and use the common parent of all files as the container.

Reply:
I had the same idea, then I thought; we have to overcomplicate the backend logic for a feature which we won't use that frequently anyway, but let's do.